lib/PPI/Node.pm

```diff
@@ -732,14 +751,18 @@ sub __replace_child {
 	my $self = shift;
 	my $key  = refaddr shift;
 	my $p    = List::Util::first {
-	         refaddr $self->{children}[$_] == $key
-	         } 0..$#{$self->{children}};
+			 refaddr $self->{children}[$_] == $key
+			 } 0..$#{$self->{children}};
+
+	return if !defined $p;
+
 	foreach ( @_ ) {
 		Scalar::Util::weaken(
 			$_PARENT{refaddr $_} = $self
 			);
 	}
 	splice( @{$self->{children}}, $p, 1, @_ );
+	delete $_PARENT{$key};
```

Comment:
@reneeb do you recall why this was done?

Reply:
`%_PARENT` is exported by `PPI::Singletons`.

Reply:
I *think* this was done to be consistent with `remove_child` as a replacement is something like remove the child and add a new child at that position.

Reply:
I think this whole section deserves to have these comments added:

```
# cache parent of new children
# replace old child with new children
# uncache parent of old child
```

Optimally the former and latter should be methods because i see that action repeated elsewhere, and its code is damn hard to read.

Reply:
> Optimally the former and latter should be methods because i see that action repeated elsewhere, and its code is damn hard to read.

For sure. I think that could happen in a different PR.